src/main/scala/analysis/RegToMemAnalysis.scala

```diff
@@ -0,0 +1,67 @@
+package analysis
+
+import ir.{MemoryLoad, *}
+import analysis.solvers.*
+
+import scala.collection.immutable
+
+/**
+ * Collects all the memory loads and the expressions that are assigned to a register but cannot be evaluated.
+ *
+ * Tracks:
+ * R_x = MemoryLoad[Base + Offset]
+ * R_x = Base + Offset
+ *
+ * Both in which constant propagation mark as TOP which is not useful.
+ */
+trait RegionAccessesAnalysis(cfg: ProgramCfg, constantProp: Map[CfgNode, Map[Variable, FlatElement[BitVecLiteral]]]) {
```

Comment:
If the issue is with constant propagation not providing useful results, maybe an analysis that can handle both constants and memory accesses should have been implemented instead of needing multiple separate analyses that are later combined?

Reply:
Constant propagation is meant for tracking constants so it is not proper to mix types with expressions

A new analysis that can handle both may not be necessarily useful because not all subsequent analyses make use of the expression result that the analysis will provide, may lead to confusion on the result of the analysis and in some instances approximating the result to Top could be more beneficial that returning unevaluated expression.

If you think merging it could be better then I can work on that

Reply:
I don't think merging it should be a priority but that sort of thing is something we should think about - considering how do all the analyses we have fit together since it's a bit convoluted at present

Reply:
Will have to add it as an issue for it then [here](https://github.com/UQ-PAC/bil-to-boogie-translator/issues/192)